### PR TITLE
Add `qe-tools` to registry with support for QE PW and QE CP input files

### PR DESCRIPTION
**Pull request by edan-bainglass**
Adding `qe-tools` to the registry. For now, it will support QE's PW and CP input files. Others may follow depending on how this goes.

A few notes:

1. Had to add an `extract` function as a middle layer between MaRDA and `qe-tools` classes. Open PR pending review. For now, the extractor schema added to the registry points to my `qe-tools` fork. Let's see how that goes.

2. The `command` for my extractor is `qe_tools.extractors.extract({{ input_path }}, {{ input_type }})`, where `input_path` is expected to be the path to the file, and `input_type` is `PW` or `CP`. I defined `input_type` under each `supported_filetypes` entry in the schema. Not sure if this is the way.

### CODE REVIEW

File: .gitattributes
Comment (ml-evs) [inline .gitattributes:7]: Does this work @PeterKraus? I thought lfs was more magic than this (and GH normally clearly delinates LFS files in PRs) We can probably just amend the PR directly if not (no need for @edan-bainglass to worry)

**Review by ml-evs — COMMENTED** [review]
Here's what I get trying to run this programmatically:]

```python
$ python
Python 3.11.7 (main, Dec 13 2023, 22:06:27) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from marda_extractors_api import extract
>>> extract("https://raw.githubusercontent.com/marda-alliance/metadata_extractors_registry/4e04310205ded7fe6ce7b6cb3a6b9b61895164ce/marda_registry/data/lfs/qe-pw-in/pw.in", "qe-pw-in", install=True, use_venv=True, extractor_definition={"id": "qe-tools", "usage": [{"method": "python", "setup": "qe_tools", "command": "qe_tools.extractors.extract({{ input_path }}, {{ input_type }})"}], "installation": [{"method": "pip", "requires_python": ">=3.8", "packages": ["qe-tools"]}], "supported_filetypes": [{"id": "qe-pw-in"}]})
Attempting to install qe-tools
Requirement already satisfied: qe-tools in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (2.1.0)
Requirement already satisfied: numpy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools) (1.26.3)
Requirement already satisfied: scipy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools) (1.11.4)
Requirement already satisfied: packaging in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools) (23.2)

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: /home/mevans/repos/marda/metadata_extractors_api/marda-venvs/env-qe-tools/bin/python -m pip install --upgrade pip
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'qe_tools' has no attribute 'extractors'
/home/mevans/.pyenv/versions/3.11.7/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mevans/repos/marda/metadata_extractors_api/marda_extractors_api/__init__.py", line 138, in extract
    return extractor.execute(
           ^^^^^^^^^^^^^^^^^^
  File "/home/mevans/repos/marda/metadata_extractors_api/marda_extractors_api/__init__.py", line 279, in execute
    output = self._execute_python_venv(command, setup)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mevans/repos/marda/metadata_extractors_api/marda_extractors_api/__init__.py", line 352, in _execute_python_venv
    subprocess.check_output(
  File "/home/mevans/.pyenv/versions/3.11.7/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mevans/.pyenv/versions/3.11.7/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/mevans/repos/marda/metadata_extractors_api/marda-venvs/env-qe-tools/bin/python', '-c', 'print("Launching extractor"); import qe_tools; import pickle; import multiprocessing.shared_memory;shm = multiprocessing.shared_memory.SharedMemory(name=\'psm_57075983\');data = pickle.dumps(qe_tools.extractors.extract(\'/tmp/tmp1ye2f1sn\', \'qe-pw-in\')); shm.buf[:len(data)] = data; print(\'Done!\');']' returned non-zero exit status 1.
```

Not sure if this is a bug at our end or with the qe-tools integration yet.

File: marda_registry/data/extractors/quantumespresso.yml
Comment (ml-evs) [inline marda_registry/data/extractors/quantumespresso.yml:17]: maybe add a lower pin here for when marda support was added?

```suggestion
          - "qe-tools > 2.1" 
```

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:17]: Reverting to git reference (`aiidateam`'s) until official release

**Comment by edan-bainglass** [comment]
I force pushed a correction - installation now points to `aiidateam/qe-tools`. Will update pointer to pypi once officially released, at which point, I'll tag minimal version in reqs.

File: marda_registry/data/extractors/quantumespresso.yml
Comment (ml-evs) [inline marda_registry/data/extractors/quantumespresso.yml:12]: Running into one more issue with the git version, that I cannot access `qe_tools.extractors` with a simple `import qe_tools`. Works fine if you directly `import qe_tools.extractors`. Guess this can be fixed at either end, as you prefer

```suggestion
      setup: qe_tools.extractors
```

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:12]: I'll fix it on my end

**Comment by edan-bainglass** [comment]
@ml-evs okay, fixed. Had to add the extractors module to the root `__init__.py`. For now, I reverted the redirect back to my repo. Let's make sure it works before I push any further changes to `qe-tools`.

File: .gitattributes
Comment (PeterKraus) [reply .gitattributes:7]: I cannot remember. But this github docs page seems to say it should be enough: https://docs.github.com/en/repositories/working-with-files/managing-large-files/configuring-git-large-file-storage

**Review by PeterKraus — COMMENTED** [review]
Looks good. I think we can finally approve it once the extraction works using the API.

File: marda_registry/data/extractors/quantumespresso.yml
Comment (ml-evs) [reply marda_registry/data/extractors/quantumespresso.yml:12]: New error:

```
ValueError: Supported parsers: ['PW', 'CP']; given: 'QE-PW-IN'
```

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:12]: I think maybe I'm misunderstanding how this is executed. The command is `qe_tools.extractos.extract({{ input_path }}, {{ input_type }})`. Under `supported_filetypes`, I specify `input_type` to be "PW" or "CP". "QE-PW-IN" suggests the `id` is passed instead of the type?

Comment (ml-evs) [reply marda_registry/data/extractors/quantumespresso.yml:12]: Whoops, this was my bad, sorry to bother you on a Sunday, I was missing your `template` instructions from my minimal example. This is something that should work but doesn't seem to yet, I think because the API assumes that variables defined in `template` won't override the real templated values (e.g., you could define `my_input_type` and use that in your usage string and it should work). I'll try to fix this next week

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:12]: I'm clearly working on a Sunday 😅

Comment (ml-evs) [reply marda_registry/data/extractors/quantumespresso.yml:12]: End-to-end example works nicely now I've actually implemented the templates...

```python
$ python
Python 3.11.7 (main, Dec 13 2023, 22:06:27) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from marda_extractors_api import extract
>>> extract("https://raw.githubusercontent.com/marda-alliance/metadata_extractors_registry/4e04310205ded7fe6ce7b6cb3a6b9b61895164ce/marda_registry/data/lfs/qe-pw-in/pw.in", "qe-pw-in", install=True, use_venv=True, extractor_definition={"id": "qe-tools", "usage": [{"method": "python", "setup": "qe_tools", "command": "qe_tools.extractors.extract({{ input_path }}, {{ input_type }})"}], "installation": [{"method": "pip", "requires_python": ">=3.8", "packages": ["git+https://github.com/aiidateam/qe-tools.git"]}], "supported_filetypes": [{"id": "qe-pw-in", "template": {"input_type": "PW"}}]})
Attempting to install qe-tools
Collecting git+https://github.com/aiidateam/qe-tools.git
  Cloning https://github.com/aiidateam/qe-tools.git to /tmp/pip-req-build-qxh8krw_
  Running command git clone --filter=blob:none --quiet https://github.com/aiidateam/qe-tools.git /tmp/pip-req-build-qxh8krw_
  Resolved https://github.com/aiidateam/qe-tools.git to commit 8c50b4c6e203b1ccea7dd90f384c2e365e74dd5c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (1.26.3)
Requirement already satisfied: scipy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (1.11.4)
Requirement already satisfied: packaging in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (23.2)

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: /home/mevans/repos/marda/metadata_extractors_api/marda-venvs/env-qe-tools/bin/python -m pip install --upgrade pip
/home/mevans/.pyenv/versions/3.11.7/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
{'structure': {'positions': [[0.005238854365041, 0.005238854365041, 0.0], [0.9697172347411749, -0.11869444788673698, -5.2917720858999996e-05], [-0.11869444788673698, 0.9697172347411749, 0.00010583544171799999]], 'species': {'names': ['O', 'H'], 'masses': [16.0, 1.0], 'pseudo_file_names': ['O.blyp-mt.UPF', 'H.blyp-vbc.UPF']}, 'cell': [[6.350126503079999, 0.0, 0.0], [0.0, 6.350126503079999, 0.0], [0.0, 0.0, 6.350126503079999]], 'atom_names': ['O', 'H', 'H']}, 'k-points': {'type': 'gamma'}}
```

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:12]: Not sure what's all that about a memory leak, but good? 👍?

Comment (ml-evs) [reply marda_registry/data/extractors/quantumespresso.yml:12]: :shushing_face:

Comment (edan-bainglass) [reply marda_registry/data/extractors/quantumespresso.yml:12]: 🤐

**Review by ml-evs — APPROVED** [review]
As this is all working now, I don't see any reason not to merge at this stage and update again once qe-tools is released on PyPI! I'll give you the final word though @edan-bainglass!

**Review by PeterKraus — APPROVED** [review]

**Comment by ml-evs** [comment]
I'm going to merge this now because I want to test some other stuff based on it; we can always revise in a future PR.

**Comment by edan-bainglass** [comment]
Sorry @ml-evs, busy bee 🐝 Thanks for the merge 🙏 I'll open a follow-up PR once qe-tools side is settled 👍

File: .gitattributes
Comment (ml-evs) [reply .gitattributes:7]: It did not work -- after merging when I pull the repo I get a warning about files not being pointers. Will correct now.

**Pull request merged**
